File: go/statement.go
// Copyright (c) 2025 ADBC Drivers Contributors
//
// This file has been modified from its original version, which is
// under the Apache License:
//
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package databricks

import (
	"context"
	"database/sql"
	"database/sql/driver"
	"errors"

	"github.com/adbc-drivers/driverbase-go/driverbase"
	"github.com/apache/arrow-adbc/go/adbc"
	"github.com/apache/arrow-go/v18/arrow"
	"github.com/apache/arrow-go/v18/arrow/array"
)

type statementImpl struct {
	driverbase.StatementImplBase
	conn              *connectionImpl
	query             string
	prepared          *sql.Stmt
	boundStream       array.RecordReader
	bulkIngestOptions driverbase.BulkIngestOptions
}

func (s *statementImpl) Close() error {
	if s.conn == nil {
		return s.ErrorHelper.Errorf(adbc.StatusInvalidState, "statement already closed")
	}
	if s.boundStream != nil {
		s.boundStream.Release()
		s.boundStream = nil
	}
	if s.prepared != nil {
		if err := s.prepared.Close(); err != nil {
			return err
		}
		s.prepared = nil
	}
	s.conn = nil
	return nil
}

func (s *statementImpl) SetOption(key, val string) error {
	if handled, err := s.bulkIngestOptions.SetOption(&s.ErrorHelper, key, val); err != nil {
		return err
	} else if handled {
		return nil
	}

	return s.ErrorHelper.Errorf(adbc.StatusNotImplemented, "unsupported statement option: %s=%s", key, val)
}

func (s *statementImpl) SetSqlQuery(query string) error {
	s.query = query
	// Reset prepared statement if query changes
	if s.prepared != nil {
		if err := s.prepared.Close(); err != nil {
			return s.ErrorHelper.Errorf(adbc.StatusInvalidState, "failed to close previous prepared statement: %v", err)
		}
		s.prepared = nil
	}
	return nil
}

func (s *statementImpl) Prepare(ctx context.Context) error {
	if s.query == "" {
		return s.ErrorHelper.Errorf(adbc.StatusInvalidState, "no query set")
	}

	stmt, err := s.conn.conn.PrepareContext(ctx, s.query)
	if err != nil {
		return s.ErrorHelper.Errorf(adbc.StatusInvalidState, "failed to prepare statement: %v", err)
	}

	s.prepared = stmt
	return nil
}

func (s *statementImpl) ExecuteQuery(ctx context.Context) (array.RecordReader, int64, error) {
	if s.boundStream != nil {
		return nil, -1, s.ErrorHelper.Errorf(adbc.StatusNotImplemented, "parameterized queries not yet implemented")
	}

	if s.query == "" {
		return nil, -1, s.ErrorHelper.Errorf(adbc.StatusInvalidState, "no query set")
	}

	// Execute query using raw driver interface to get Arrow batches
	// This works for both prepared and unprepared statements since
	// databricks-sql-go doesn't do server-side preparation
	var driverRows driver.Rows
	var err error
	err = s.conn.conn.Raw(func(driverConn interface{}) error {
		// Use raw driver interface for direct Arrow access
		queryerCtx := driverConn.(driver.QueryerContext)
		var driverArgs []driver.NamedValue
		driverRows, err = queryerCtx.QueryContext(ctx, s.query, driverArgs)
		return err
	})

	if err != nil {
		return nil, -1, s.ErrorHelper.Errorf(adbc.StatusInternal, "failed to execute query: %v", err)
	}

	defer func() {
		if driverRows == nil {
			return
		}
		if closeErr := driverRows.Close(); closeErr != nil {
			err = errors.Join(err, closeErr)
		}
	}()

	// Use the IPC stream interface (zero-copy)
	reader, err := newIPCReaderAdapter(ctx, driverRows, s.conn.useArrowNativeGeospatial)
	if err != nil {
		return nil, -1, s.ErrorHelper.Errorf(adbc.StatusInternal, "failed to create IPC reader adapter: %v", err)
	}
	driverRows = nil // Prevent double close in defer

	// Return -1 for rowsAffected (unknown) since we can't count without consuming
	// The ADBC spec allows -1 to indicate "unknown number of rows affected"
	return reader, -1, nil
}

func (s *statementImpl) ExecuteUpdate(ctx context.Context) (int64, error) {
	if s.bulkIngestOptions.IsSet() {
		return s.executeIngest(ctx)
	}

	if s.boundStream != nil {
		return -1, s.ErrorHelper.Errorf(adbc.StatusInvalidState, "bound data provided but no ingest target set")
	}

	var result sql.Result
	var err error

	if s.prepared != nil {
		result, err = s.prepared.ExecContext(ctx)
	} else if s.query != "" {
		result, err = s.conn.conn.ExecContext(ctx, s.query)
	} else {
		return -1, s.ErrorHelper.Errorf(adbc.StatusInvalidState, "no query set")
	}

	if err != nil {
		return -1, s.ErrorHelper.Errorf(adbc.StatusInternal, "failed to execute update: %v", err)
	}

	rowsAffected, err := result.RowsAffected()
	if err != nil {
		return -1, s.ErrorHelper.Errorf(adbc.StatusInternal, "failed to get rows affected: %v", err)
	}

	return rowsAffected, nil
}

func (s *statementImpl) Bind(ctx context.Context, values arrow.RecordBatch) error {
	if s.boundStream != nil {
		s.boundStream.Release()
	}
	stream, err := array.NewRecordReader(values.Schema(), []arrow.RecordBatch{values})
	if err != nil {
		return s.ErrorHelper.Errorf(adbc.StatusInternal, "failed to create record reader")
	}
	s.boundStream = stream
	return nil
}

func (s *statementImpl) BindStream(ctx context.Context, stream array.RecordReader) error {
	if s.boundStream != nil {
		s.boundStream.Release()
	}
	stream.Retain()
	s.boundStream = stream
	return nil
}

func (s *statementImpl) GetParameterSchema() (*arrow.Schema, error) {
	// This would require parsing the SQL query to determine parameter types
	// For now, return nil to indicate unknown schema
	return nil, s.ErrorHelper.Errorf(adbc.StatusNotImplemented, "parameter schema detection not implemented")
}

func (s *statementImpl) SetSubstraitPlan(plan []byte) error {
	// Databricks SQL doesn't support Substrait plans
	return s.ErrorHelper.Errorf(adbc.StatusNotImplemented, "Substrait plans not supported")
}

func (s *statementImpl) ExecutePartitions(ctx context.Context) (*arrow.Schema, adbc.Partitions, int64, error) {
	// Databricks SQL doesn't support partitioned result sets
	return nil, adbc.Partitions{}, -1, s.ErrorHelper.Errorf(adbc.StatusNotImplemented, "partitioned result sets not supported")
}

File: rust/src/statement.rs
// Copyright (c) 2025 ADBC Drivers Contributors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Statement implementation for the Databricks ADBC driver.

use crate::client::DatabricksClient;
use crate::error::DatabricksErrorHelper;
use crate::reader::ResultReaderAdapter;
use crate::types::sea::ExecuteParams;
use adbc_core::error::Result;
use adbc_core::options::{OptionStatement, OptionValue};
use adbc_core::Optionable;
use arrow_array::RecordBatchReader;
use arrow_schema::Schema;
use driverbase::error::ErrorHelper;
use std::sync::Arc;
use tokio::runtime::Handle as RuntimeHandle;
use tracing::debug;

/// Represents a SQL statement that can be executed against Databricks.
///
/// A Statement is created from a Connection and is used to execute SQL
/// queries and retrieve results.
#[derive(Debug)]
pub struct Statement {
    /// The SQL query to execute.
    query: Option<String>,
    /// Databricks client for executing queries.
    client: Arc<dyn DatabricksClient>,
    /// Session ID for this statement's connection.
    session_id: String,
    /// Tokio runtime handle for async operations.
    runtime_handle: RuntimeHandle,
    /// Current statement ID (set after execution).
    current_statement_id: Option<String>,
    /// Convert Databricks geometry structs to geoarrow.wkb.
    use_arrow_native_geospatial: bool,
}

impl Statement {
    /// Creates a new Statement.
    pub(crate) fn new(
        client: Arc<dyn DatabricksClient>,
        session_id: String,
        runtime_handle: RuntimeHandle,
        use_arrow_native_geospatial: bool,
    ) -> Self {
        Self {
            query: None,
            client,
            session_id,
            runtime_handle,
            current_statement_id: None,
            use_arrow_native_geospatial,
        }
    }

    /// Returns the current SQL query.
    pub fn sql_query(&self) -> Option<&str> {
        self.query.as_deref()
    }
}

impl Optionable for Statement {
    type Option = OptionStatement;

    fn set_option(&mut self, key: Self::Option, _value: OptionValue) -> Result<()> {
        Err(DatabricksErrorHelper::set_unknown_option(&key).to_adbc())
    }

    fn get_option_string(&self, key: Self::Option) -> Result<String> {
        Err(DatabricksErrorHelper::get_unknown_option(&key).to_adbc())
    }

    fn get_option_bytes(&self, key: Self::Option) -> Result<Vec<u8>> {
        Err(DatabricksErrorHelper::get_unknown_option(&key).to_adbc())
    }

    fn get_option_int(&self, key: Self::Option) -> Result<i64> {
        Err(DatabricksErrorHelper::get_unknown_option(&key).to_adbc())
    }

    fn get_option_double(&self, key: Self::Option) -> Result<f64> {
        Err(DatabricksErrorHelper::get_unknown_option(&key).to_adbc())
    }
}

impl adbc_core::Statement for Statement {
    fn set_sql_query(&mut self, query: impl AsRef<str>) -> Result<()> {
        self.query = Some(query.as_ref().to_string());
        Ok(())
    }

    fn set_substrait_plan(&mut self, _plan: impl AsRef<[u8]>) -> Result<()> {
        Err(DatabricksErrorHelper::not_implemented()
            .message("Substrait plans")
            .to_adbc())
    }

    fn prepare(&mut self) -> Result<()> {
        // Databricks doesn't have server-side prepared statements, but
        // ADBC consumers (e.g. DuckDB's adbc_scanner) call prepare before
        // execute. Accept it as a no-op if a query has been set.
        if self.query.is_none() {
            return Err(DatabricksErrorHelper::invalid_state()
                .message("No query set before prepare")
                .to_adbc());
        }
        Ok(())
    }

    fn get_parameter_schema(&self) -> Result<Schema> {
        Err(DatabricksErrorHelper::not_implemented()
            .message("get_parameter_schema")
            .to_adbc())
    }

    fn bind(&mut self, _batch: arrow_array::RecordBatch) -> Result<()> {
        Err(DatabricksErrorHelper::not_implemented()
            .message("bind parameters")
            .to_adbc())
    }

    fn bind_stream(&mut self, _stream: Box<dyn RecordBatchReader + Send>) -> Result<()> {
        Err(DatabricksErrorHelper::not_implemented()
            .message("bind_stream")
            .to_adbc())
    }

    fn execute(&mut self) -> Result<impl RecordBatchReader + Send> {
        let query = self.query.as_ref().ok_or_else(|| {
            DatabricksErrorHelper::invalid_state()
                .message("No query set")
                .to_adbc()
        })?;

        debug!("Executing query: {}", query);

        // Execute statement via DatabricksClient (handles polling and reader creation)
        let result = self
            .runtime_handle
            .block_on(self.client.execute_statement(
                &self.session_id,
                query,
                &ExecuteParams::default(),
            ))
            .map_err(|e| e.to_adbc())?;

        // Store statement ID for cancellation/cleanup
        self.current_statement_id = Some(result.statement_id);

        // Wrap in adapter for RecordBatchReader trait
        ResultReaderAdapter::new(
            result.reader,
            result.manifest.as_ref(),
            self.use_arrow_native_geospatial,
        )
        .map_err(|e| e.to_adbc())
    }

    fn execute_update(&mut self) -> Result<Option<i64>> {
        // Execute and count affected rows
        let reader = self.execute()?;

        // For UPDATE/INSERT/DELETE, there are no result rows
        // but the manifest might have row count info
        // Drain the reader (shouldn't have data for DML)
        for _batch in reader {
            // Consume batches
        }

        // TODO: Extract affected row count from response manifest
        Ok(None)
    }

    fn execute_schema(&mut self) -> Result<Schema> {
        // Execute and get schema from reader
        let reader = self.execute()?;
        Ok((*reader.schema()).clone())
    }

    fn execute_partitions(&mut self) -> Result<adbc_core::PartitionedResult> {
        Err(DatabricksErrorHelper::not_implemented()
            .message("execute_partitions")
            .to_adbc())
    }

    fn cancel(&mut self) -> Result<()> {
        if let Some(ref statement_id) = self.current_statement_id {
            debug!("Canceling statement: {}", statement_id);
            self.runtime_handle
                .block_on(self.client.cancel_statement(statement_id))
                .map_err(|e| e.to_adbc())?;
        }
        Ok(())
    }
}

impl Drop for Statement {
    fn drop(&mut self) {
        // Clean up statement resources
        if let Some(ref statement_id) = self.current_statement_id {
            let _ = self
                .runtime_handle
                .block_on(self.client.close_statement(statement_id));
        }
    }
}

#[cfg(test)]
mod tests {
    // Note: Full statement tests require mock DatabricksClient
    // Integration tests should be added in a separate test module
}
